### PR TITLE
Initial release candidate

### PR DESCRIPTION
**Pull request by yugensoft**

### CODE REVIEW

**Comment by Xon** [comment]
FYI, you should put any php version check into addon.json via;
```
    "require" : {
        "php": ["5.6.0", "PHP 5.6.0+"]
    },
```

This way your installer can use those language features.
Note; I've moving to a minimum of php5.6+ for all my add-ons.

File: upload/src/addons/SV/TitleEditHistory/EditHistoryHandler/Thread.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	/**
+	 * @param \SV\TitleEditHistory\XF\Entity\Thread|Entity $content
+	 * @param \XF\Entity\EditHistory $history
+	 * @param \XF\Entity\EditHistory|null $previous
+	 *
+	 * @return mixed
+	 */
+	public function revertToVersion(Entity $content, \XF\Entity\EditHistory $history, \XF\Entity\EditHistory $previous = null) {
```
Comment (Xon) [inline upload/src/addons/SV/TitleEditHistory/EditHistoryHandler/Thread.php]: `revertToVersion` should `XF\Service\Thread\Editor` similar to how `\XF\EditHistory\Post` does

File: upload/src/addons/SV/TitleEditHistory/XF/Entity/Thread.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+		if (\XF::visitor()->hasNodePermission($this->node_id, 'manageAnyThread'))
+		{
+			return true;
+		}
+
+		return false;
+	}
+
+	protected function _isNotInsideSilentEditWindow() {
```
Comment (Xon) [inline upload/src/addons/SV/TitleEditHistory/XF/Entity/Thread.php]: The edit history logic should be moved into `XF\Service\Thread\Editor` similar to `XF\Service\Post\Editor::setMessage`, `XF\Service\Post\Editor::setMessage::setupEditHistory`.

File: upload/src/addons/SV/TitleEditHistory/XF/Entity/Thread.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+		if ($this->isUpdate() && $this->isChanged('title') && $this->_isNotInsideSilentEditWindow())
+		{
+			$this->_insertTitleEditHistory();
+		}
+
+		return $upstream;
+	}
+
+	protected function _insertTitleEditHistory()
```
Comment (Xon) [inline upload/src/addons/SV/TitleEditHistory/XF/Entity/Thread.php]: Once moved to the Thread Editor, this should be replaced with;
```
/** @var \XF\Repository\EditHistory $repo */
$repo = $this->repository('XF:EditHistory');
$repo->insertEditHistory('thread', $thread, $visitor, $this->oldTitle, $this->app->request()->getIp());
```

File: upload/src/addons/SV/TitleEditHistory/_output/content_type_fields/thread_title-edit_history_handler_class.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+    "content_type": "thread_title",
+    "field_name": "edit_history_handler_class",
+    "field_value": "SV\\TitleEditHistory\\EditHistoryHandler\\Thread"
```
Comment (Xon) [inline upload/src/addons/SV/TitleEditHistory/_output/content_type_fields/thread_title-edit_history_handler_class.json]: Rename from EditHistoryHandler => EditHistory, and thankfully it doesn't require a custom upgrade step (unlike XF1)

**Review by Xon — CHANGES_REQUESTED** [review]

**Comment by Xon** [comment]
Note; `\XF\Service\Thread\Editor::_save()` currently has some horrible bugs with respect to transaction handling.

Add the insert edit history logic around a `parent::_save()` call with a transaction. Since all _save() does is call `$thread->save()`

**Comment by yugensoft** [comment]
Alright let me know if that's what you had in mind.

This edithistory & 'service' system seemed like an awful lot of object-oriented abstraction to accomplish simply viewing a previous state. Do you know why Xenforo Co decided to head even deeper down the OO path with XF2, and didn't move more down the functional programming & immutable data route like Facebook etc?

**Comment by Xon** [comment]
LGTM

Keep in mind that XF1 is dated 2008-2010 for it's major architecture design, and XF2 was 2015-2016 for it's initial design/implementation. XenForo is very conservative on baseline requirements and performance. Maybe XF3 in 2028 or so will be a different story.

**Pull request merged**